packages/itwin/tree-widget/src/components/trees/stateless/common/UseNodeHighlighting.tsx

```diff
@@ -0,0 +1,204 @@
+    const info = state.current.nodeInfoMap.get(nodeId);
+    if (!info) {
+      return undefined;
+    }
+    if (activeIndex === undefined) {
+      return { chunks: info.matches };
+    }
+    const isActive = info && activeIndex >= info.startIndex && activeIndex < info.startIndex + info.matches.length;
+    return isActive ? { activeChunkIndex: activeIndex - info.startIndex, chunks: info.matches } : { chunks: info.matches };
+  }, []);
+
+  const computeHighlightState = useCallback(() => {
+    const newState: HighlightState = { nodeInfoMap: new Map(), totalMatches: 0 };
+    let activeNodeInfo: { nodeId: string; info: NodeChunkInfo } | undefined;
+
+    const computeHighlightStateRecursively = (nodes: Array<PresentationTreeNode>) => {
+      nodes.forEach((node) => {
+        if (!isPresentationHierarchyNode(node)) {
+          return;
+        }
+
+        // keep track of previous active chunk
+        const nodeInfo = getNodeChunkInfo(node.id, activeMatchIndexRef.current);
+        if (nodeInfo?.activeChunkIndex !== undefined) {
+          activeNodeInfo = { nodeId: node.id, info: nodeInfo };
+        }
+
+        const matches = findChunks(node.label, textToHighlight!, caseSensitive ?? false);
+        newState.nodeInfoMap.set(node.id, { startIndex: newState.totalMatches, matches });
+        newState.totalMatches += matches.length;
+
+        if (typeof node.children !== "boolean") {
+          computeHighlightStateRecursively(node.children);
+        }
+      });
+    };
+
+    computeHighlightStateRecursively(rootNodes!);
+
+    let newActiveIndex = activeMatchIndexRef.current ?? 0;
+    if (activeNodeInfo && newActiveIndex !== 0) {
+      // update active index to not cause active chunk jumps when hierarchy changes
+      const updatedInfo = newState.nodeInfoMap.get(activeNodeInfo.nodeId);
+      updatedInfo && (newActiveIndex = updatedInfo?.startIndex + activeNodeInfo.info.activeChunkIndex!);
+    }
+    state.current = newState;
+    onHighlightChanged?.(newActiveIndex, newState.totalMatches);
+  }, [rootNodes, textToHighlight, caseSensitive, getNodeChunkInfo, onHighlightChanged]);
+
+  useEffect(() => {
+    if (rootNodes && textToHighlight) {
+      computeHighlightState();
+      return;
+    }
+    state.current = { nodeInfoMap: new Map(), totalMatches: 0 };
+  }, [rootNodes, textToHighlight, caseSensitive, computeHighlightState]);
+
+  useEffect(() => {
+    onHighlightChanged?.(0, state.current.totalMatches);
+  }, [textToHighlight, caseSensitive, onHighlightChanged]);
+
+  useEffect(() => {
+    // focus on currently highlighted node to scroll it into view
+    for (const nodeId of state.current.nodeInfoMap.keys()) {
+      if (getNodeChunkInfo(nodeId, activeMatchIndex)?.activeChunkIndex !== undefined) {
+        const nodeElement = document.querySelector(`[id="${nodeId}"]`) as HTMLElement;
+        nodeElement?.focus();
+      }
+    }
+  }, [activeMatchIndex, getNodeChunkInfo]);
```

Comment:
This probably could be more generic. I think `useNodeHighlighting` should find and return `nodeId` and leave scrolling to element up to the tree renderer.